### PR TITLE
[feat] Add CFI checks to the driver library

### PR DESCRIPTION
**Pull request by mhatrevi**

### CODE REVIEW

File: drivers/src/ecc384.rs [hunk 1]
```diff
@@ -254,7 +255,7 @@ impl Ecc384 {
         };
 
         // Pairwise consistency check.
-        if perform_pct {
+        if cfi_launder(priv_key_usage.ecc_private_key()) {
```
Comment (korran) [inline drivers/src/ecc384.rs:258]: I think this branch is too dangerous. Can we just remove the branch and always validate? Why would anyone want to generate a private key that can't be used as a private key?

File: drivers/src/ecc384.rs [hunk 2]
```diff
@@ -336,12 +339,13 @@ impl Ecc384 {
     ) -> CaliptraResult<Ecc384Signature> {
         let mut sig_result = self.sign_internal(priv_key, data, trng);
         let sig = okmutref(&mut sig_result)?;
-        match self.verify(pub_key, data, sig)? {
+        let result = self.verify(pub_key, data, sig)?;
```
Comment (korran) [inline drivers/src/ecc384.rs:342]: Add `// TODO(kor): unrolled CFI comparison of verify result`

**Review by korran — COMMENTED** [review]
Should we be wrapping these functions with cfi_impl_fn?

File: drivers/src/ecc384.rs (hunk 1 above)
Comment (FerralCoder) [reply drivers/src/ecc384.rs:258]: I agree, I can't think of any valid reason why this PCT should ever be skipped.

Comment (mhatrevi) [reply drivers/src/ecc384.rs:258]: Discussed offline; we'll block creating a key-pair if [ecc_pkey_dest_valid](https://chipsalliance.github.io/caliptra-rtl/main/internal-regs/?p=clp.ecc_reg.ecc_kv_wr_pkey_ctrl#ecc_pkey_dest_valid.desc) is not set

**Comment by mhatrevi** [comment]
> Should we be wrapping these functions with cfi_impl_fn?

Added.

File: drivers/src/sha256.rs [hunk 3]
```diff
@@ -33,6 +35,8 @@ pub trait Sha256Alg {
 
     fn digest_init(&mut self) -> CaliptraResult<Self::DigestOp<'_>>;
     fn digest(&mut self, buf: &[u8]) -> CaliptraResult<Array4x8>;
+    #[cfg(not(feature = "no-cfi"))]
+    fn __cfi_digest(&mut self, buf: &[u8]) -> CaliptraResult<Array4x8>;
```
Comment (korran) [inline drivers/src/sha256.rs:39]: Hmm, this is pretty messy. I wonder if we could use nested functions instead.

Comment (korran) [reply drivers/src/sha256.rs:39]: As is, we need to update other users of this trait (in the fuzz tests)

Comment (korran) [reply drivers/src/sha256.rs:39]: Here's a fairly nice solution:

https://github.com/chipsalliance/caliptra-sw/pull/1060

Probably the sweet spot of complexity vs code size is the second commit in that chain. The third makes it 3x more complex just to save 36 bytes (making the entire PR increase the size by just 4 bytes).

Comment (mhatrevi) [reply drivers/src/sha256.rs]: Discussed offline; we'll decorate the Crypto::sha256_digest function with CFI instead of the Sha256::digest function.

**Review by korran — DISMISSED** [review]

**Review by korran — APPROVED** [review]

**Review by jhand2 — APPROVED** [review]

**Review by FerralCoder — APPROVED** [review]

**Pull request merged**